[FFL-2385] Add Stale Flags concept page (#37279)
DataDog/documentation: content/en/feature_flags/concepts/stale_flags.md

---
title: Stale Flags

> description: 'Identify and clean up stale feature flags to reduce flag debt'

Reviewer: Description fields should end with a period:

```suggestion
description: 'Identify and clean up stale feature flags to reduce flag debt.'
```

## How Datadog detects stale flags

A flag is evaluated for staleness only if it is not archived, not marked as permanent, and not tied to a running experiment. A flag is marked `STALE` when any of the following signals are true:

- **Fully rolled out**: The same variant has been served to 100% of traffic in all production environments for the past 30 days. This indicates the flag is behaving like static code and may no longer provide value as a runtime control.
- **No evaluations**: No evaluation events have been recorded for the flag in the past 30 days across all environments. This may indicate that the flag is no longer referenced in your application code.
- **No activity**: The flag has not been modified in the past 30 days across all environments. This may indicate that the rollout has been deprioritized or completed.

**Note**: Evaluation logging is not supported by every SDK, and some customers opt out. As a result, the **No evaluations** signal is a strong but imperfect indicator that a flag is ready to be archived.

## View stale flags in the flag list

> The [Feature Flags list page][1] shows a {{< ui >}}STALE{{< /ui >}} chip next to each flag that Datadog has identified as stale. Hover the chip to see which signal triggered it.

Reviewer: Minor: `Hover over the chip` is the standard phrasing.

```suggestion
The [Feature Flags list page][1] shows a {{< ui >}}STALE{{< /ui >}} chip next to each flag that Datadog has identified as stale. Hover over the chip to see which signal triggered it.
```

- Use the {{< ui >}}Staleness{{< /ui >}} filter to narrow the list to stale flags only.
- Filter by team to see the flag debt your team is responsible for.

> {{< img src="feature_flags/stale_flag_chip.png" alt="Feature flag list view showing the STALE chip and staleness filter" style="width:100%;" >}}

Reviewer: Alt text should end with a period (Datadog accessibility standard).

```suggestion
{{< img src="feature_flags/stale_flag_chip.png" alt="Feature flag list view showing the STALE chip and staleness filter." style="width:100%;" >}}
```

## Review a stale flag

> When you open a stale flag, a banner at the top of the flag details page explains why the flag is considered stale and recommends actions to resolve it. The banner highlights one **primary suggestion** based on the highest-confidence signal and the lowest-risk action, with additional secondary actions available alongside it.

Reviewer: The trailing `with ... available alongside it` reads as an afterthought participial. Claude's suggestion — make it a clean clause:

```suggestion
When you open a stale flag, a banner at the top of the flag details page explains why the flag is considered stale and recommends actions to resolve it. The banner highlights one **primary suggestion** based on the highest-confidence signal and the lowest-risk action, and provides additional secondary actions.
```

> {{< img src="feature_flags/stale_flag_banner.png" alt="Stale banner on the feature flag details page with recommended actions" style="width:100%;" >}}

Reviewer: Alt text should end with a period.

```suggestion
{{< img src="feature_flags/stale_flag_banner.png" alt="Stale banner on the feature flag details page with recommended actions." style="width:100%;" >}}
```

## Manage stale flags

> ### Generate a removal PR with Bits AI

Reviewer: The heading says "Bits AI," but the button label below (and the modal alt text) read "Generate Removal PR with **Bits**." Worth confirming the actual UI label and aligning the heading with it — if the feature is branded "Bits AI," keep it; if the button literally reads "Bits," match that.

From the stale banner, select {{< ui >}}Generate Removal PR with Bits{{< /ui >}} to open a pull request that removes the flag-gating logic from your application code. This is the recommended action when the flag is fully rolled out and is still receiving evaluation traffic.

**Note**: Remove the flag from your code and deploy the change before you archive the flag. Archiving a flag that is still referenced in code can cause unexpected behavior in your application.

> {{< img src="feature_flags/generate_removal_pr.png" alt="Generate Removal PR with Bits modal" style="width:100%;" >}}

Reviewer: Alt text should end with a period.

```suggestion
{{< img src="feature_flags/generate_removal_pr.png" alt="Generate Removal PR with Bits modal." style="width:100%;" >}}
```

### Archive the flag

Archiving marks the flag as inactive so it is no longer evaluated or editable. Archive a flag after you have removed its references from application code and confirmed it is no longer receiving evaluation traffic. Archiving a flag requires approval and cancels any experiments running on the flag.

### Mark as permanent

> Some flags are intentionally long-lived, such as killswitches or permission gates. Select {{< ui >}}Mark as Permanent{{< /ui >}} on the stale banner to designate a flag as a permanent operational control and exclude it from future stale detection. Permanent flags show a {{< ui >}}PERMANENT{{< /ui >}} indicator in the flag details sidebar.

Reviewer: `kill switches` is two words (per Merriam-Webster):

```suggestion
Some flags are intentionally long-lived, such as kill switches or permission gates. Select {{< ui >}}Mark as Permanent{{< /ui >}} on the stale banner to designate a flag as a permanent operational control and exclude it from future stale detection. Permanent flags show a {{< ui >}}PERMANENT{{< /ui >}} indicator in the flag details sidebar.
```

> {{< img src="feature_flags/mark_as_permanent.png" alt="Feature flag marked as PERMANENT on the flag details page" style="width:100%;" >}}

Reviewer: Alt text should end with a period.

```suggestion
{{< img src="feature_flags/mark_as_permanent.png" alt="Feature flag marked as PERMANENT on the flag details page." style="width:100%;" >}}
```

### Dismiss for 30 days

If a flag is temporarily inactive or cleanup is intentionally deferred, select {{< ui >}}Dismiss for 30 days{{< /ui >}} to silence the staleness indicator. The flag is re-evaluated for staleness after 30 days, so you do not need to manually reverse the dismissal.

## Permissions

The `FeatureFlagConfigWrite` permission is required to mark a flag as permanent. Archiving a flag also requires `FeatureFlagConfigWrite`.

## Further reading

{{< partial name="whats-next/whats-next.html" >}}

> [1]: https://app.datadoghq.com/feature-flags

Reviewer: Couldn't verify this link from here (the page is auth-gated). Worth confirming `/feature-flags` resolves — and note that a hardcoded `app.datadoghq.com` link sends EU/US3/etc. users to the US1 site rather than their own.

Author: Seems most of the other pages do this though? At least according to claude running over it all. Is this a new convention?

Reviewer: ah this comment was a mistake. sorry about that. yeah it's extremely common! my freaking Claude did this lol
